### PR TITLE
Handle empty S3 payloads (#4514)

### LINKED ISSUE #4514: object_store: Uploading empty file to S3 results in "411 Length Required"
**Describe the bug**

When using the S3 implementation of `object_store` a `put()` call with an empty content fails on the server side with a "411 Length Required" error:

> Generic S3 error: response error "\<?xml version="1.0" encoding="UTF-8"?>
\<Error>\<Code>MissingContentLength\</Code>\<Message>You must provide the Content-Length HTTP header.\</Message>\</Error>", after 0 retries: HTTP status client error (411 Length Required) for url (https://s3.us-west-2.amazonaws.com/...)

**To Reproduce**

```rs
s3_store.put(&"foo".into(), Bytes::new()).await.unwrap()
```

**Expected behavior**

The call succeeds and uploads an empty file to S3.

**Additional context**

- https://github.com/rust-lang/crates.io/issues/6793

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #4514

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Review by alamb — COMMENTED** [review]
The code looks good to me -- thank you @tustvold 

I am just verifying the test coverage... Hang tight

File: object_store/src/lib.rs [hunk 1]
```diff
@@ -1258,6 +1258,15 @@ mod tests {
         }
 
         delete_fixtures(storage).await;
+
+        let path = Path::from("empty");
```
Comment (alamb) [inline object_store/src/lib.rs:1262]: This test appears to pass even without the code changes in this PR. See https://github.com/apache/arrow-rs/pull/4519 for example

**Review by alamb — COMMENTED** [review]
The code looks good, but I am not sure about the test

**Comment by tustvold** [comment]
You need to test against real S3, the emulator doesn't seem to exhibit the same issue

**Comment by alamb** [comment]
> You need to test against real S3, the emulator doesn't seem to exhibit the same issue

I will try

**Comment by alamb** [comment]
@tustvold  can you give me the command you used run the tests against a real S3 bucket (obviously I will supply my own AWS credentials)

I got this far (but then I stumbled on what the "metadata endpoint" should be

```
TEST_INTEGRATION=1 OBJECT_STORE_BUCKET=foo  cargo test -p object_store --features=aws
```

**Comment by tustvold** [comment]
Just omit the endpoint, note also that the credential names should be prefixed by OBJECT_STORE

**Review by alamb — APPROVED** [review]
Ok, I ran the new test against an actual S3 bucket and without the code changes in this PR it fails like this:

```

---- aws::tests::s3_test stdout ----
thread 'aws::tests::s3_test' panicked at 'called `Result::unwrap()` on an `Err` value: Generic { store: "S3", source: Error { retries: 0, message: "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>MissingContentLength</Code><Message>You must provide the Content-Length HTTP header.</Message><RequestId>....</RequestId><HostId>....</HostId></Error>", source: Some(reqwest::Error { kind: Status(411), url: Url { scheme: "https", cannot_be_a_base: false, username: "", password: None, host: Some(Domain("s3.us-east-1.amazonaws.com")), port: None, path: "/alamb-test-bucket/empty", query: None, fragment: None } }), status: Some(411) } }', src/lib.rs:1263:48
```

With this PR it passes

Nice work @tustvold  -- thank you

**Pull request merged**